File: packages/protocol/src/commands.ts
/**
 * Command palette related types.
 * 
 * Note: These match Rust structs exactly (snake_case) since we don't
 * configure serde to rename to camelCase.
 */

/** A searchable command shown in the command palette. */
export interface SearchResult {
  /** Unique identifier (extension_id:command_id) */
  id: string;
  /** Extension that provides this command */
  extension_id: string;
  /** Command ID within the extension */
  command_id: string;
  /** Display name */
  name: string;
  /** Optional description */
  description?: string;
  /** Optional icon identifier */
  icon?: string;
  /** Search relevance score (higher = more relevant) */
  score?: number;
}

/** Request to search for commands. */
export interface SearchRequest {
  /** The search query */
  query: string;
  /** Maximum number of results to return */
  limit?: number;
}

/** Request to execute a command. */
export interface ExecuteRequest {
  /** Extension ID */
  extension_id: string;
  /** Command ID within the extension */
  command_id: string;
  /** Optional input/prompt */
  input?: string;
}

/** Result of command execution - success variant. */
export interface ExecuteResultSuccess {
  type: "success";
  /** Optional output data */
  output?: unknown;
}

/** Result of command execution - process variant. */
export interface ExecuteResultProcess {
  type: "process";
  /** Process/session ID for tracking */
  session_id: string;
}

/** Result of command execution - error variant. */
export interface ExecuteResultError {
  type: "error";
  /** Error message */
  message: string;
}

/** Result of command execution. */
export type ExecuteResult =
  | ExecuteResultSuccess
  | ExecuteResultProcess
  | ExecuteResultError;

File: crates/protocol/src/commands.rs
//! Command palette related types.

use serde::{Deserialize, Serialize};

/// A searchable command shown in the command palette.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SearchResult {
    /// Unique identifier (extension_id:command_id)
    pub id: String,
    /// Extension that provides this command
    pub extension_id: String,
    /// Command ID within the extension
    pub command_id: String,
    /// Display name
    pub name: String,
    /// Optional description
    #[serde(default)]
    pub description: Option<String>,
    /// Optional icon identifier
    #[serde(default)]
    pub icon: Option<String>,
    /// Search relevance score (higher = more relevant)
    #[serde(default)]
    pub score: f64,
}

/// Request to search for commands.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SearchRequest {
    /// The search query
    pub query: String,
    /// Maximum number of results to return
    #[serde(default = "default_limit")]
    pub limit: usize,
}

fn default_limit() -> usize {
    20
}

/// Request to execute a command.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ExecuteRequest {
    /// Extension ID
    pub extension_id: String,
    /// Command ID within the extension
    pub command_id: String,
    /// Optional input/prompt
    #[serde(default)]
    pub input: Option<String>,
}

/// Result of command execution.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ExecuteResult {
    /// Command completed successfully
    Success {
        /// Optional output data
        #[serde(default)]
        output: Option<serde_json::Value>,
    },
    /// Command spawned a long-running process
    Process {
        /// Process/session ID for tracking
        session_id: String,
    },
    /// Command failed
    Error {
        /// Error message
        message: String,
    },
}
